### ISSUE TITLE
Request for Guidance: Node Label Coverage Issue in Dataset Processing for Node Classification Task

### ISSUE THREAD

**Issue by qmy0701** [yule-BUAA/DyGLib#12]
In your dataset processing, each edge is associated with an edge label, but for the node classification task, what we need is the node label. I noticed there was a previous issue (https://github.com/yule-BUAA/DyGLib/issues/3) where you mentioned that the label of each edge is actually the label of the source node. However, there are some issues with my dataset:

For the node classification dataset I'm using, I traversed all edges, adding the source node and destination node of each edge to two separate lists. After tallying, I performed a deduplication operation on both lists by using the set() function, and found that the sizes of the source set and destination set are both smaller than the total number of nodes. So, associating an edge label (which is essentially the label of the source node) with each edge doesn't completely cover the labels of all my nodes. Could you please advise on how to resolve this issue within your code framework?

**Comment by qmy0701** [comment]
Hello. Could you offer any help on this issue?

Thanks in advance!

**Comment by yule-BUAA** [comment]
Hi,

I want to know in your mentioned case, does either the source node or the destination node have a label at each timestamp? Or the source nodes or the destination nodes only have labels at particular timestamps?

**Comment by qmy0701** [comment]
Node label corresponds to a particular timestamp.

**Comment by yule-BUAA** [comment]
If I understand correctly, your case is similar to the node classification task in [TGB](https://tgb.complexdatalab.com/docs/leader_nodeprop/), where nodes have labels at particular timestamps. I have established the [DyGLib_TGB](https://github.com/yule-BUAA/DyGLib_TGB) repository to support this.

The overall process of node classification is still based on the link prediction pipeline but with several modifications:
1. I added the `interact_types` property to mark each interaction with values "train", "validate", "test" or "just_update", where the "just_update" interactions mean interacted nodes have no labels but padded zero labels. See [here](https://github.com/yule-BUAA/DyGLib_TGB/blob/fc2af7e09e132133a1e45602927ffedb69b6ab1b/utils/DataLoader.py#L60) and [here](https://github.com/yule-BUAA/DyGLib_TGB/blob/fc2af7e09e132133a1e45602927ffedb69b6ab1b/utils/DataLoader.py#L227)
2. During the model computation, I performed backpropagation using edge labels (which are the node labels actually) with edges with "train" `interact_types`, and evaluated edges with "validate" and "test" `interact_types`. Edges with "just_update" are either only used to update the memory modules of memory-based models like TGN, or ignored by models without memory modules. See [here](https://github.com/yule-BUAA/DyGLib_TGB/blob/fc2af7e09e132133a1e45602927ffedb69b6ab1b/train_node_classification.py#L165) and [here](https://github.com/yule-BUAA/DyGLib_TGB/blob/fc2af7e09e132133a1e45602927ffedb69b6ab1b/evaluate_models_utils.py#L213).

The above modifications enable the support of node classification tasks with minor changes to the link prediction pipeline.

For your mentioned "So, associating an edge label (which is essentially the label of the source node) with each edge doesn't completely cover the labels of all my nodes", do you mean that both source and destination nodes will have labels? If that is the case, you can simply expand the [node label matrix](https://github.com/yule-BUAA/DyGLib_TGB/blob/fc2af7e09e132133a1e45602927ffedb69b6ab1b/utils/DataLoader.py#L226) to contain both node labels. Or do you mean that sometimes the label belongs to the source node and sometimes it belongs to the destination node? If so, you can add a flag to each edge to indicate where the label comes from (i.e., source node or destination node), and use the flag to find the corresponding node when computing loss.

I hope I understand your questions correctly, and if this is not the case, please let me know. 

Feel free to ask if you still have any questions.

**Comment by yule-BUAA** [comment]
Close this issue now.

Feel free to reopen it when there are any further questions.

**Comment by mgao97** [comment]
Hello,

I have a question regarding the setup for node classification within a dynamic and directed graph scenario.

In this scenario, both the source and destination nodes remain consistent across all timestamps. The existing edge labels correspond to the labels of the source nodes at each timestamp. However, these labels do not account for the labels of the destination nodes, even though the destination nodes are included in our training and testing datasets.

Could you please advise on how to construct the dataset to accommodate node classification tasks under these conditions?

Any suggestions would be greatly appreciated.
